lib/languages/js/source-code/source-code.js

```diff
@@ -58,20 +58,24 @@ const CODE_PATH_EVENTS = [
  * @private
  */
 function validate(ast) {
+    if (!ast) {
+        throw new TypeError(`Unexpected empty AST. (${ast})`);
+    }
+
     if (!ast.tokens) {
-        throw new Error("AST is missing the tokens array.");
+        throw new TypeError("AST is missing the tokens array.");
```

Comment:
Following https://github.com/eslint/eslint/pull/19458#discussion_r1970354879, I went ahead and updated the rest of the error classes too.